frame/support/src/storage/types/counted_nmap.rs

```diff
@@ -0,0 +1,1427 @@
+		if removed_value.is_some() {
+			CounterFor::<Prefix>::mutate(|value| value.saturating_dec());
+		}
+		QueryKind::from_optional_value_to_query(removed_value)
+	}
+
+	/// Swap the values of two key-pairs.
+	pub fn swap<KOther, KArg1, KArg2>(key1: KArg1, key2: KArg2)
+	where
+		KOther: KeyGenerator,
+		KArg1: EncodeLikeTuple<Key::KArg> + TupleToEncodedIter,
+		KArg2: EncodeLikeTuple<KOther::KArg> + TupleToEncodedIter,
+	{
+		<Self as MapWrapper>::Map::swap::<KOther, _, _>(key1, key2)
+	}
+
+	/// Store a value to be associated with the given keys from the map.
+	pub fn insert<KArg, VArg>(key: KArg, val: VArg)
+	where
+		KArg: EncodeLikeTuple<Key::KArg> + TupleToEncodedIter + Clone,
+		VArg: EncodeLike<Value>,
+	{
+		if !<Self as MapWrapper>::Map::contains_key(key.clone()) {
+			CounterFor::<Prefix>::mutate(|value| value.saturating_inc());
+		}
+		<Self as MapWrapper>::Map::insert(key, val)
+	}
+
+	/// Remove the value under the given keys.
+	pub fn remove<KArg: EncodeLikeTuple<Key::KArg> + TupleToEncodedIter + Clone>(key: KArg) {
+		if <Self as MapWrapper>::Map::contains_key(key.clone()) {
+			CounterFor::<Prefix>::mutate(|value| value.saturating_dec());
+		}
+		<Self as MapWrapper>::Map::remove(key)
+	}
+
+	/// Attempt to remove items from the map matching a `partial_key` prefix.
+	///
+	/// Returns [`MultiRemovalResults`](sp_io::MultiRemovalResults) to inform about the result. Once
+	/// the resultant `maybe_cursor` field is `None`, then no further items remain to be deleted.
+	///
+	/// NOTE: After the initial call for any given map, it is important that no further items
+	/// are inserted into the map which match the `partial key`. If so, then the map may not be
+	/// empty when the resultant `maybe_cursor` is `None`.
+	///
+	/// # Limit
+	///
+	/// A `limit` must be provided in order to cap the maximum
+	/// amount of deletions done in a single call. This is one fewer than the
+	/// maximum number of backend iterations which may be done by this operation and as such
+	/// represents the maximum number of backend deletions which may happen. A `limit` of zero
+	/// implies that no keys will be deleted, though there may be a single iteration done.
+	///
+	/// # Cursor
+	///
+	/// A *cursor* may be passed in to this operation with `maybe_cursor`. `None` should only be
+	/// passed once (in the initial call) for any given storage map and `partial_key`. Subsequent
+	/// calls operating on the same map/`partial_key` should always pass `Some`, and this should be
+	/// equal to the previous call result's `maybe_cursor` field.
+	pub fn clear_prefix<KP: Clone>(
+		partial_key: KP,
+		limit: u32,
+		maybe_cursor: Option<&[u8]>,
+	) -> sp_io::MultiRemovalResults
+	where
+		Key: HasKeyPrefix<KP>,
+	{
+		let mut on_remove = 0;
+		Self::iter_prefix_values(partial_key.clone()).for_each(|_| {
+			on_remove += 1;
+		});
+		CounterFor::<Prefix>::mutate(|value| {
+			*value = value.saturating_sub(on_remove);
+		});
```

Comment:
how is `limit` being respected here?

Reply:
What really should happen is that the `MultiRemovalResults` from `clear_prefix` should be looked into to check and see what the actual number of deleted items are; we then update the counter accordingly.

Reply:
updated